### ISSUE TITLE
Alternative way of 'local'

### ISSUE THREAD

**Issue by vuxyp** [koalaman/shellcheck#1681]
Hi,

```sh
local NEW_VAR1="ABC"
```

gives me this warning: 

> In POSIX sh, 'local' is undefined.

I call many times my custom functions from shell. I have local variables inside these functions. If I don't use "local" keyword for these variables they are still using by other functions which is not make sense.

What should I do here? What is the best practice.

Example:

```sh
test(){
    local NEW_VAR1="ABC1"
    NEW_VAR2="ABC2"
}
```

Open the command line interpreter:
```sh
test
echo $NEW_VAR1

echo $NEW_VAR2
ABC2
```

You see? NEW_VAR1 and NEW_VAR2 are private variables for test() function. It should not be seen to others.

I want my functions to work bash and zsh on all operating systems as possible.

Thanks you

**Comment by Nightfirecat** [comment]
`local` is allowed within bash, you just need to specify it in the shebang.

```bash
#!/bin/bash
foo() {
  local foo="bar"
  echo "$foo"
}
```

**Comment by vuxyp** [comment]
@Nightfirecat below command starts bash in POSIX compatible mode:

> bash --posix

and it works properly with "local" variables. I try it now on my machine. I am confused :( 

Also what do you suggest for my script to work with POSIX standard mode of bash? If we don't use local, than where is no way to support private variables. right?

**Comment by koalaman** [comment]
@vuxyp The job of `bash --posix` is to run POSIX compatible scripts, not to fail POSIX-incompatible ones.

POSIX explicitly says the behavior of `local` is undefined, so Bash can do whatever it wants with it and still be compatible. It's up to the script to not depend on undefined behavior.

You're right that strictly POSIX shell scripts can't define local variables, so you can either pick a naming scheme where global variables used only in single functions won't collide, or try to control the scope with subshells.

PS: `local` is a very common extension that works on bash, zsh, ksh, dash and BusyBox, so it wouldn't be totally unreasonable for you to just choose to depend on it anyways
